### ISSUE TITLE
is_sparse doesn't work for lgCMatrix objects

### ISSUE THREAD

**Issue by LTLA** [Bioconductor/DelayedArray#63]
```r
library(DelayedArray)
Y <- Matrix::rsparsematrix(100, 100, 0.1)
is_sparse(Y) # TRUE
is_sparse(Y > 0) # FALSE
```

The same problem applies to other sparse formats:

```r
is_sparse(Z <- as(Y, "dgTMatrix")) # FALSE
is_sparse(Z > 0) # FALSE
```

Though one might be forgiven for not supporting the ultra-rare `dgRMatrix`:

```r
AA <- as(as.matrix(Y), "dgRMatrix")
is_sparse(AA) # FALSE
is_sparse(AA > 0) # FALSE 
```

<details>
<summary>Session info</summary>

```
R Under development (unstable) (2020-03-11 r77927)
Platform: x86_64-apple-darwin17.7.0 (64-bit)
Running under: macOS High Sierra 10.13.6

Matrix products: default
BLAS:   /Users/luna/Software/R/trunk/lib/libRblas.dylib
LAPACK: /Users/luna/Software/R/trunk/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets
[8] methods   base

other attached packages:
[1] DelayedArray_0.15.1 IRanges_2.23.4      S4Vectors_0.27.5
[4] BiocGenerics_0.35.2 matrixStats_0.56.0

loaded via a namespace (and not attached):
[1] compiler_4.0.0  Matrix_1.2-18   tools_4.0.0     grid_4.0.0
[5] lattice_0.20-41
```
</details>

**Comment by hpages** [comment]
Yeah only dgCMatrix objects are supported by `is_sparse()` / `extract_sparse_array()` / `read_sparse_block()` at the moment. For other sparse formats, we currently get false negatives. What this means concretely is that that blocks from dgTMatrix or dgRMatrix objects cannot be read with `read_sparse_block()` right now, only with `read_block()`. I could add support for dgTMatrix and dgRMatrix which means implementing `extract_sparse_array()` methods for these objects. What other sparseMatrix subclasses should we support? Even though we should probably support them all in theory, I'd like to take a more pragmatic approach for now.

Note that `is_sparse()` and `extract_sparse_array()` methods are strongly correlated. They're defined for the same objects:
```
> showMethods("is_sparse")
Function: is_sparse (package DelayedArray)
x="ANY"
x="DelayedAbind"
x="DelayedAperm"
x="DelayedNaryIsoOp"
x="DelayedSubassign"
x="DelayedSubset"
x="DelayedUnaryIsoOp"
x="DelayedUnaryIsoOpStack"
x="DelayedUnaryIsoOpWithArgs"
x="dgCMatrix"
x="SparseArraySeed"

> showMethods("extract_sparse_array")
Function: extract_sparse_array (package DelayedArray)
x="DelayedAbind"
x="DelayedAperm"
x="DelayedNaryIsoOp"
x="DelayedSubassign"
x="DelayedSubset"
x="DelayedUnaryIsoOp"
x="DelayedUnaryIsoOpStack"
x="DelayedUnaryIsoOpWithArgs"
x="dgCMatrix"
x="SparseArraySeed"
```
This is because the purpose of `is_sparse()` is to provide a switch from dense block processing (i.e. `read_block()`-based) to sparse block processing (i.e. `read_sparse_block()`-based). See `DelayedArray:::BLOCK_write_to_sink()` for a typical example. The various `extract_sparse_array()` methods are the workhorses behind `read_sparse_block()`.

**Comment by LTLA** [comment]
Probably just `lgCMatrix` for now, the others are rare enough to not be a major priority.

**Comment by hpages** [comment]
This is implemented in DelayedArray 0.15.2:
```
library(DelayedArray)
Y <- Matrix::rsparsematrix(100, 100, 0.1) > 0

is_sparse(Y)
# [1] TRUE

extract_sparse_array(Y, list(1:10, 1:10))
# An object of class "SparseArraySeed"
# Slot "dim":
# [1] 10 10
#
# Slot "nzindex":
#       [,1] [,2]
#  [1,]    8    2
#  [2,]    8    3
#  [3,]    9    3
#  [4,]    9    5
#  [5,]    2    6
#  [6,]    5    6
#  [7,]    3    7
#  [8,]    6    8
#  [9,]    4    9
# [10,]    7    9
# [11,]    9    9
# [12,]    3   10
#
# Slot "nzdata":
#  [1]  TRUE FALSE  TRUE  TRUE  TRUE FALSE  TRUE FALSE FALSE FALSE  TRUE  TRUE

as(extract_sparse_array(Y, list(1:10, 1:10)), "sparseMatrix")
# 10 x 10 sparse Matrix of class "lgCMatrix"
#                        
#  [1,] . . . . . . . . . .
#  [2,] . . . . . | . . . .
#  [3,] . . . . . . | . . |
#  [4,] . . . . . . . . : .
#  [5,] . . . . . : . . . .
#  [6,] . . . . . . . : . .
#  [7,] . . . . . . . . : .
#  [8,] . | : . . . . . . .
#  [9,] . . | . | . . . | .
# [10,] . . . . . . . . . .
```
